src/dragonball/src/dbs_pci/src/msi.rs

```diff
@@ -0,0 +1,759 @@
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 3 => {
+                    data[0] = (self.mask_bits >> 24) as u8;
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 4 => {
+                    self.get_pending_state(data, 0);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 5 => {
+                    self.get_pending_state(data, 1);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 6 => {
+                    self.get_pending_state(data, 2);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 7 => {
+                    self.get_pending_state(data, 3);
+                }
+                _ => debug!("invalid offset"),
+            },
+            2 => match offset {
+                0 => LittleEndian::write_u16(data, self.cap_id_next),
+                MSI_MSG_CTL_OFFSET => LittleEndian::write_u16(data, self.msg_ctl),
+                x if x == msg_data_offset => {
+                    LittleEndian::write_u16(data, self.msg_data);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                    LittleEndian::write_u16(data, self.mask_bits as u16);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 2 => {
+                    LittleEndian::write_u16(data, (self.mask_bits >> 16) as u16);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 4 => {
+                    self.get_pending_state(data, 0);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 6 => {
+                    self.get_pending_state(data, 2);
+                }
+                _ => debug!("invalid offset"),
+            },
+            4 => match offset {
+                0x0 => LittleEndian::write_u32(
+                    data,
+                    self.cap_id_next as u32 | ((self.msg_ctl as u32) << 16),
+                ),
+                MSI_MSG_ADDR_LO_OFFSET => LittleEndian::write_u32(data, self.msg_addr_lo),
+                x if addr_hi_offset.is_some() && x == addr_hi_offset.unwrap() => {
+                    LittleEndian::write_u32(data, self.msg_addr_hi);
+                }
+                x if x == msg_data_offset => {
+                    LittleEndian::write_u32(data, self.msg_data as u32);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                    LittleEndian::write_u32(data, self.mask_bits);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 4 => {
+                    self.get_pending_state(data, 0);
+                }
```

Comment:
Ditto.

Reply:
this is for writing bits according to pci spec. I have added a summarised comment above
"Support writing mask_bits, MSI control information, message data, message address."